element_interface/utils.py

```diff
@@ -140,31 +147,98 @@ def ingest_csv_to_table(
-    """Return value for key in a nested dictionary
-
-    Search through a nested dictionary for a key and returns its value.  If there are
-    more than one key with the same name at different depths, the algorithm returns the
-    value of the least nested key.
+def str_to_bool(value) -> bool:
+    """Return whether the provided string represents true. Otherwise false.
 
     Args:
-        key (str): Key used to search through a nested dictionary
-        dictionary (dict): Nested dictionary
+        value (any): Any input
 
     Returns:
-        value (any): value of the input argument `key`
+        bool (bool): True if value in ("y", "yes", "t", "true", "on", "1")
     """
-    if key in dictionary:
-        return dictionary[key]
-    for value in dictionary.values():
-        if isinstance(value, dict):
-            a = recursive_search(key, value)
-            if a is not None:
-                return a
-    return None
+    # Due to distutils equivalent depreciation in 3.10
+    # Adopted from github.com/PostHog/posthog/blob/master/posthog/utils.py
+    if not value:
+        return False
+    return str(value).lower() in ("y", "yes", "t", "true", "on", "1")
+
+
+def write_csv(content: list, path: pathlib.PosixPath):
+    """General function for writing strings to lines in CSV
+
+    Args
+        content (list): list of strings, each as row of CSV
+        path (pathlib.PosixPath): where to wriite new CSV
+    """
+    with open(path, "w") as f:
+        for line in content:
+            f.write(line + "\n")
+
+
+class QuietStdOut:
+    """Context for quieting standard output, and setting datajoint loglevel to warning
+
+    Used in pytest functions to render cleaer output showing only pass/fail
+
+    Example:
+        with QuietStdOut:
+            table.delete(safemode=False)
+
+    """
+
+    def __enter__(self):
+        os.environ["DJ_LOG_LEVEL"] = "WARNING"
+        self._original_stdout = sys.stdout
+        sys.stdout = open(os.devnull, "w")
+
+    def __exit__(self):  # , exc_type, exc_val, exc_tb):
+        os.environ["DJ_LOG_LEVEL"] = "INFO"
+        sys.stdout.close()
+        sys.stdout = self._original_stdout
+
+
+def insert1_skip_full_duplicates(table: Table, entry: dict):
+    """Insert one entry into a table, ignoring duplicates if all entries match.
+
+    Duplicates on either primary or secondary attributes log existing entries. After
+    validation, this functionality will be integrated into core DataJoint.
+
+    Cases:
+        0. New entry: insert as normal
+        1. Entry has an exact match: return silently
+        2. Same primary key, new secondary attributes: log warning with full entry
+        3. New primary key, same secondary attributes: log warning with existing key
+
+    Arguments:
+        table (Table): datajoint Table object entry (dict): table entry as a dictionary
+    """
+    if table & entry:  # Test for 1. Return silently if exact match
+        return
```

Comment:
Do not test before inserting since this is racy, i.e. it will lead to errors in race conditions where two workers will attempt the same operation.